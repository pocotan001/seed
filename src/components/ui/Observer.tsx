import { pick } from "lodash";
import * as React from "react";

type RenderNode = (isIntersecting: boolean) => React.ReactNode;

interface ObserverProps extends IntersectionObserverInit {
  children: RenderNode | React.ReactNode;
  once?: boolean;
  onEnter?: (
    entry: IntersectionObserverEntry,
    observer: IntersectionObserver
  ) => void;
  onLeave?: (
    entry: IntersectionObserverEntry,
    observer: IntersectionObserver
  ) => void;
}

interface ObserverState {
  isIntersecting: boolean;
}

/**
 * Acting as a wrapper for the `IntersectionObserver` API
 */
export default class Observer extends React.PureComponent<
  ObserverProps,
  ObserverState
> {
  state = {
    isIntersecting: false
  };

  el = React.createRef<Element>();
  observer!: IntersectionObserver;

  handleIntersection: IntersectionObserverCallback = (entries, observer) => {
    const { once, onEnter, onLeave } = this.props;
    const [entry] = entries;
    const { isIntersecting } = entry;
    const hasChange = this.state.isIntersecting !== isIntersecting;

    if (!hasChange) {
      return;
    }

    if (isIntersecting) {
      if (onEnter) {
        onEnter(entry, observer);
      }

      if (once) {
        this.observer.unobserve(this.el.current!);
      }
    } else if (onLeave) {
      onLeave(entry, observer);
    }

    this.setState({ isIntersecting });
  };

  componentDidMount() {
    const opts: IntersectionObserverInit = pick(this.props, [
      "root",
      "rootMargin",
      "threshold"
    ]);

    this.observer = new IntersectionObserver(this.handleIntersection, opts);
    this.observer.observe(this.el.current!);
  }

  componentWillUnmount() {
    this.observer.unobserve(this.el.current!);
  }

  render() {
    const { children } = this.props;
    const el =
      typeof children === "function"
        ? children(this.state.isIntersecting)
        : children;

    return React.cloneElement(React.Children.only(el), {
      ref: this.el
    });
  }
}
