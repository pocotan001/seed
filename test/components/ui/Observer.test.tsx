import * as React from "react";
import * as renderer from "react-test-renderer";
import Observer from "~/components/ui/Observer";

describe("<Observer>", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Observer>
          <div>alo</div>
        </Observer>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
