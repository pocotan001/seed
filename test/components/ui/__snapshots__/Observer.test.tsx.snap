// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Observer> should render correctly 1`] = `
<div>
  alo
</div>
`;
